src/adapter/adapterDiscovery.ts

```diff
@@ -0,0 +1,430 @@
+                        resolve([adapter, path, baudRate]);
+                    } else {
+                        reject(new Error(`Adapter ${adapter} is not supported.`));
+                    }
+                } else {
+                    bj.destroy();
+                    reject(
+                        new Error(
+                            `Coordinator returned wrong Zeroconf format! The following values are expected:\n` +
+                                `txt.radio_type, got: ${service.txt?.radio_type}\n` +
+                                `txt.baud_rate, got: ${service.txt?.baud_rate}\n` +
+                                `address, got: ${service.addresses?.[0]}\n` +
+                                `port, got: ${service.port}`,
+                        ),
+                    );
+                }
+            } else {
+                bj.destroy();
+                reject(new Error(`Coordinator [${mdnsDevice}] not found after timeout of ${mdnsTimeout}ms!`));
+            }
+        });
+    });
+}
+
+export async function findTCPAdapter(path: string, adapter?: Adapter): Promise<[adapter: ValidAdapter, path: string]> {
+    const regex = /^tcp:\/\/(?:[0-9]{1,3}\.){3}[0-9]{1,3}:\d{1,5}$/gm;
+
+    if (!regex.test(path)) {
+        throw new Error(`Invalid TCP path, expected format: tcp://<host>:<port>`);
+    }
+
+    if (!adapter || adapter === 'auto') {
+        throw new Error(`Cannot discover TCP adapters at this time. Please specify valid 'adapter' and 'path' manually.`);
+    }
+
+    return [adapter, path];
+}
+
+/**
+ * Discover adapter using mDNS, TCP or USB.
+ *
+ * @param adapter The adapter type.
+ *   - mDNS: Unused.
+ *   - TCP: Required, cannot discover at this time.
+ *   - USB: Optional, limits the discovery to the specified adapter type.
+ * @param path The path to the adapter.
+ *   - mDNS: Required, serves to initiate the discovery.
+ *   - TCP: Required, cannot discover at this time.
+ *   - USB: Optional, limits the discovery to the specified path.
+ * @returns adapter An adapter type supported by Z2M. While result is TS-typed, this should be validated against actual values before use.
+ * @returns path Path to adapter.
+ * @returns baudRate [optional] Discovered baud rate of the adapter. Valid only for mDNS discovery at the moment.
+ */
+export async function discoverAdapter(
+    adapter?: Adapter,
+    path?: string,
+): Promise<[adapter: ValidAdapter, path: string, baudRate?: number | undefined]> {
+    if (path) {
+        if (path.startsWith('mdns://')) {
+            return await findmDNSAdapter(path);
+        } else if (path.startsWith('tcp://')) {
+            return await findTCPAdapter(path, adapter);
+        } else if (adapter && adapter !== 'auto') {
+            const matched = await matchUSBAdapter(adapter, path);
+
+            /* istanbul ignore else */
+            if (!matched) {
+                logger.error(`Unable to match USB adapter: ${adapter} | ${path}`, NS);
```

Comment:
In case both adapter and path is specified, why do we need to do any matching and log an error when it doesn't? I expect that setting both adapter and path wouldn't do any auto detection at all.

Reply:
It acts as "pre-validation" of the config.
It used to do the same before, but only to pre-validate `path`.
https://github.com/Koenkk/zigbee-herdsman/blob/c661d2ea744ddd80b4d0b37c6b54dec8cb6324e3/src/adapter/adapter.ts#L170-L182

It's not "strictly" necessary.

Reply:
Then I propose to log it as debug, if auto detect doesn't work (which it doesn't with the HA addon for example) users will always see this error which can be confusing.